File: java/design_pattern/proxy/MainProxy.java
/**
 * Comprehensive demonstration of the Proxy Pattern in Java
 *
 * The Proxy pattern provides a surrogate or placeholder for another object
 * to control access to it.
 */

import java.util.HashMap;
import java.util.Map;

// Subject interface
interface Image {
    void display();
    String getFilename();
}

// Real Subject - expensive to create
class RealImage implements Image {
    private String filename;

    public RealImage(String filename) {
        this.filename = filename;
        loadFromDisk();
    }

    private void loadFromDisk() {
        System.out.println("  [RealImage] Loading image from disk: " + filename);
        // Simulate expensive operation
        try {
            Thread.sleep(100);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
    }

    @Override
    public void display() {
        System.out.println("  [RealImage] Displaying: " + filename);
    }

    @Override
    public String getFilename() {
        return filename;
    }
}

// Virtual Proxy - lazy initialization
class ProxyImage implements Image {
    private RealImage realImage;
    private String filename;

    public ProxyImage(String filename) {
        this.filename = filename;
        System.out.println("  [ProxyImage] Created proxy for: " + filename);
    }

    @Override
    public void display() {
        if (realImage == null) {
            realImage = new RealImage(filename);
        }
        realImage.display();
    }

    @Override
    public String getFilename() {
        return filename;
    }
}

// Protection Proxy example

interface Document {
    void read();
    void write(String content);
    void delete();
}

class RealDocument implements Document {
    private String name;
    private String content;

    public RealDocument(String name) {
        this.name = name;
        this.content = "Original content of " + name;
    }

    @Override
    public void read() {
        System.out.println("  [RealDocument] Reading: " + content);
    }

    @Override
    public void write(String content) {
        this.content = content;
        System.out.println("  [RealDocument] Content updated to: " + content);
    }

    @Override
    public void delete() {
        System.out.println("  [RealDocument] Document deleted: " + name);
    }
}

// Protection Proxy with access control
class ProtectedDocumentProxy implements Document {
    private RealDocument realDocument;
    private String userRole;
    private String documentName;

    public ProtectedDocumentProxy(String documentName, String userRole) {
        this.documentName = documentName;
        this.userRole = userRole;
    }

    private void initDocument() {
        if (realDocument == null) {
            realDocument = new RealDocument(documentName);
        }
    }

    @Override
    public void read() {
        initDocument();
        // Everyone can read
        realDocument.read();
    }

    @Override
    public void write(String content) {
        if (!userRole.equals("admin") && !userRole.equals("editor")) {
            System.out.println("  [ProtectionProxy] Access denied! " + userRole + " cannot write.");
            return;
        }
        initDocument();
        realDocument.write(content);
    }

    @Override
    public void delete() {
        if (!userRole.equals("admin")) {
            System.out.println("  [ProtectionProxy] Access denied! Only admin can delete.");
            return;
        }
        initDocument();
        realDocument.delete();
    }
}

// Caching Proxy example

interface DataService {
    String fetchData(String key);
}

class RealDataService implements DataService {
    @Override
    public String fetchData(String key) {
        System.out.println("  [RealDataService] Fetching data from database for key: " + key);
        // Simulate database call
        try {
            Thread.sleep(50);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        }
        return "Data for " + key + " from DB";
    }
}

class CachingProxy implements DataService {
    private RealDataService realService;
    private Map<String, String> cache = new HashMap<>();

    public CachingProxy() {
        this.realService = new RealDataService();
    }

    @Override
    public String fetchData(String key) {
        if (cache.containsKey(key)) {
            System.out.println("  [CachingProxy] Returning cached data for: " + key);
            return cache.get(key);
        }

        String data = realService.fetchData(key);
        cache.put(key, data);
        System.out.println("  [CachingProxy] Data cached for: " + key);
        return data;
    }

    public void clearCache() {
        cache.clear();
        System.out.println("  [CachingProxy] Cache cleared");
    }
}

// Logging Proxy example

interface Calculator {
    int add(int a, int b);
    int subtract(int a, int b);
    int multiply(int a, int b);
    int divide(int a, int b);
}

class RealCalculator implements Calculator {
    @Override
    public int add(int a, int b) { return a + b; }

    @Override
    public int subtract(int a, int b) { return a - b; }

    @Override
    public int multiply(int a, int b) { return a * b; }

    @Override
    public int divide(int a, int b) { return a / b; }
}

class LoggingCalculatorProxy implements Calculator {
    private RealCalculator calculator = new RealCalculator();

    @Override
    public int add(int a, int b) {
        int result = calculator.add(a, b);
        System.out.println("  [Log] add(" + a + ", " + b + ") = " + result);
        return result;
    }

    @Override
    public int subtract(int a, int b) {
        int result = calculator.subtract(a, b);
        System.out.println("  [Log] subtract(" + a + ", " + b + ") = " + result);
        return result;
    }

    @Override
    public int multiply(int a, int b) {
        int result = calculator.multiply(a, b);
        System.out.println("  [Log] multiply(" + a + ", " + b + ") = " + result);
        return result;
    }

    @Override
    public int divide(int a, int b) {
        if (b == 0) {
            System.out.println("  [Log] divide(" + a + ", " + b + ") - Error: Division by zero!");
            throw new ArithmeticException("Division by zero");
        }
        int result = calculator.divide(a, b);
        System.out.println("  [Log] divide(" + a + ", " + b + ") = " + result);
        return result;
    }
}

// Remote Proxy simulation

interface RemoteService {
    String executeQuery(String query);
}

class RemoteServiceProxy implements RemoteService {
    private String serverAddress;

    public RemoteServiceProxy(String serverAddress) {
        this.serverAddress = serverAddress;
    }

    @Override
    public String executeQuery(String query) {
        System.out.println("  [RemoteProxy] Connecting to " + serverAddress);
        System.out.println("  [RemoteProxy] Sending query: " + query);
        // Simulate network call
        System.out.println("  [RemoteProxy] Waiting for response...");
        String result = "Result from " + serverAddress + " for query: " + query;
        System.out.println("  [RemoteProxy] Received response");
        return result;
    }
}

public class MainProxy {
    public static void main(String[] args) {
        System.out.println("=== Proxy Pattern Demonstration ===\n");

        // Virtual Proxy example
        System.out.println("--- 1. Virtual Proxy (Lazy Loading) ---");
        System.out.println("Creating image proxies (no loading yet):\n");

        Image image1 = new ProxyImage("photo1.jpg");
        Image image2 = new ProxyImage("photo2.jpg");
        Image image3 = new ProxyImage("photo3.jpg");

        System.out.println("\nFirst access to image1 (loads from disk):");
        image1.display();

        System.out.println("\nSecond access to image1 (already loaded):");
        image1.display();

        System.out.println("\nFirst access to image2:");
        image2.display();

        System.out.println("\nimage3 never accessed - never loaded!\n");

        // Protection Proxy example
        System.out.println("--- 2. Protection Proxy (Access Control) ---");

        System.out.println("\nAdmin user operations:");
        Document adminDoc = new ProtectedDocumentProxy("secret.txt", "admin");
        adminDoc.read();
        adminDoc.write("New admin content");
        adminDoc.delete();

        System.out.println("\nEditor user operations:");
        Document editorDoc = new ProtectedDocumentProxy("report.txt", "editor");
        editorDoc.read();
        editorDoc.write("Editor's content");
        editorDoc.delete();  // Should be denied

        System.out.println("\nViewer user operations:");
        Document viewerDoc = new ProtectedDocumentProxy("readme.txt", "viewer");
        viewerDoc.read();
        viewerDoc.write("Unauthorized edit");  // Should be denied
        viewerDoc.delete();  // Should be denied
        System.out.println();

        // Caching Proxy example
        System.out.println("--- 3. Caching Proxy ---");
        CachingProxy dataService = new CachingProxy();

        System.out.println("\nFirst fetch (cache miss):");
        String data1 = dataService.fetchData("user_123");
        System.out.println("  Result: " + data1);

        System.out.println("\nSecond fetch same key (cache hit):");
        String data2 = dataService.fetchData("user_123");
        System.out.println("  Result: " + data2);

        System.out.println("\nFetch different key (cache miss):");
        String data3 = dataService.fetchData("user_456");
        System.out.println("  Result: " + data3);
        System.out.println();

        // Logging Proxy example
        System.out.println("--- 4. Logging Proxy ---");
        Calculator calc = new LoggingCalculatorProxy();

        calc.add(10, 5);
        calc.subtract(10, 5);
        calc.multiply(10, 5);
        calc.divide(10, 5);
        System.out.println();

        // Remote Proxy example
        System.out.println("--- 5. Remote Proxy ---");
        RemoteService remoteDb = new RemoteServiceProxy("db.example.com:5432");
        String result = remoteDb.executeQuery("SELECT * FROM users");
        System.out.println("  Result: " + result);

        System.out.println("\n=== Summary ===");
        System.out.println("Proxy pattern types:");
        System.out.println("  - Virtual Proxy: lazy initialization, load on demand");
        System.out.println("  - Protection Proxy: access control based on permissions");
        System.out.println("  - Caching Proxy: cache results of expensive operations");
        System.out.println("  - Logging Proxy: log method calls and parameters");
        System.out.println("  - Remote Proxy: represent objects in different address spaces");
        System.out.println("\nKey benefits:");
        System.out.println("  - Control access to the real object");
        System.out.println("  - Add functionality without changing real object");
        System.out.println("  - Manage lifecycle of real object");
    }
}

File: python/design_pattern/proxy/main_proxy.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""
Proxy Pattern

The Proxy pattern provides a surrogate or placeholder for another object to
control access to it. The proxy acts as an intermediary, adding functionality
like lazy loading, access control, logging, or caching.

Key Components:
- Subject: Interface for RealSubject and Proxy
- RealSubject: The actual object that proxy represents
- Proxy: Controls access to RealSubject
"""

from abc import ABC, abstractmethod
from datetime import datetime
import time


# Subject interface
class Image(ABC):
    """Abstract interface for images."""

    @abstractmethod
    def display(self) -> str:
        pass

    @abstractmethod
    def get_info(self) -> dict[str, str]:
        pass


# Real Subject
class HighResolutionImage(Image):
    """Real subject: expensive-to-load high resolution image."""

    def __init__(self, filename: str):
        self.filename = filename
        self._load_image()

    def _load_image(self) -> None:
        """Simulate expensive image loading."""
        print(f"  [RealImage] Loading '{self.filename}' from disk...")
        time.sleep(0.1)  # Simulate loading time
        self._data = f"[Image data for {self.filename}]"
        self._width = 3840
        self._height = 2160
        print(f"  [RealImage] Loaded {self._width}x{self._height} image")

    def display(self) -> str:
        return f"Displaying {self.filename} ({self._width}x{self._height})"

    def get_info(self) -> dict[str, str]:
        return {
            "filename": self.filename,
            "resolution": f"{self._width}x{self._height}",
            "size": f"{len(self._data)} bytes",
        }


# Virtual Proxy (Lazy Loading)
class ImageProxy(Image):
    """Virtual proxy that delays loading until actually needed."""

    def __init__(self, filename: str):
        self.filename = filename
        self._real_image: HighResolutionImage | None = None

    def _get_real_image(self) -> HighResolutionImage:
        """Lazy load the real image."""
        if self._real_image is None:
            print("  [Proxy] First access - loading image")
            self._real_image = HighResolutionImage(self.filename)
        return self._real_image

    def display(self) -> str:
        return self._get_real_image().display()

    def get_info(self) -> dict[str, str]:
        return self._get_real_image().get_info()


# Protection Proxy (Access Control)
class SecureImage(Image):
    """Protection proxy that controls access based on permissions."""

    def __init__(self, filename: str, allowed_users: list[str]):
        self.filename = filename
        self._allowed_users = allowed_users
        self._real_image: HighResolutionImage | None = None
        self._current_user: str = ""

    def set_user(self, username: str) -> None:
        """Set current user for access control."""
        self._current_user = username

    def _check_access(self) -> bool:
        """Check if current user has access."""
        if self._current_user in self._allowed_users:
            print(f"  [SecureProxy] Access granted to {self._current_user}")
            return True
        else:
            print(f"  [SecureProxy] Access DENIED to {self._current_user}")
            return False

    def _get_real_image(self) -> HighResolutionImage | None:
        if self._real_image is None:
            self._real_image = HighResolutionImage(self.filename)
        return self._real_image

    def display(self) -> str:
        if not self._check_access():
            return f"ACCESS DENIED: User '{self._current_user}' cannot view {self.filename}"
        img = self._get_real_image()
        return img.display() if img else "Error loading image"

    def get_info(self) -> dict[str, str]:
        if not self._check_access():
            return {"error": "Access denied"}
        img = self._get_real_image()
        return img.get_info() if img else {}


# Logging Proxy
class LoggingImageProxy(Image):
    """Logging proxy that records all access to the image."""

    def __init__(self, filename: str):
        self.filename = filename
        self._real_image: HighResolutionImage | None = None
        self._access_log: list[dict[str, str]] = []

    def _log(self, action: str) -> None:
        entry = {
            "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            "action": action,
            "filename": self.filename,
        }
        self._access_log.append(entry)
        print(f"  [LogProxy] {entry['timestamp']} - {action}")

    def _get_real_image(self) -> HighResolutionImage:
        if self._real_image is None:
            self._log("LOAD")
            self._real_image = HighResolutionImage(self.filename)
        return self._real_image

    def display(self) -> str:
        self._log("DISPLAY")
        return self._get_real_image().display()

    def get_info(self) -> dict[str, str]:
        self._log("GET_INFO")
        return self._get_real_image().get_info()

    def get_access_log(self) -> list[dict[str, str]]:
        return self._access_log.copy()


# Caching Proxy
class CachingImageProxy(Image):
    """Caching proxy that caches expensive operation results."""

    _cache: dict[str, HighResolutionImage] = {}

    def __init__(self, filename: str):
        self.filename = filename

    def _get_real_image(self) -> HighResolutionImage:
        if self.filename not in self._cache:
            print(f"  [CacheProxy] Cache MISS for {self.filename}")
            self._cache[self.filename] = HighResolutionImage(self.filename)
        else:
            print(f"  [CacheProxy] Cache HIT for {self.filename}")
        return self._cache[self.filename]

    def display(self) -> str:
        return self._get_real_image().display()

    def get_info(self) -> dict[str, str]:
        return self._get_real_image().get_info()

    @classmethod
    def get_cache_size(cls) -> int:
        return len(cls._cache)

    @classmethod
    def clear_cache(cls) -> None:
        cls._cache.clear()


# Remote Proxy (simulated)
class RemoteImageProxy(Image):
    """Simulates remote proxy for accessing images on remote server."""

    def __init__(self, filename: str, server: str = "cdn.example.com"):
        self.filename = filename
        self.server = server
        self._cached_data: dict[str, str] | None = None

    def display(self) -> str:
        print(f"  [RemoteProxy] Fetching {self.filename} from {self.server}...")
        time.sleep(0.05)  # Simulate network latency
        return f"Displaying remote image: https://{self.server}/{self.filename}"

    def get_info(self) -> dict[str, str]:
        if self._cached_data is None:
            print(f"  [RemoteProxy] Requesting metadata from {self.server}...")
            time.sleep(0.05)
            self._cached_data = {
                "filename": self.filename,
                "server": self.server,
                "url": f"https://{self.server}/{self.filename}",
            }
        return self._cached_data


def main() -> None:
    print("=" * 60)
    print("Proxy Pattern - Image Loading Demo")
    print("=" * 60)

    # Demo 1: Virtual Proxy (Lazy Loading)
    print("\n--- Virtual Proxy (Lazy Loading) ---")
    print("Creating proxy (no loading yet):")
    proxy = ImageProxy("vacation_photo.jpg")
    print("Proxy created.\n")

    print("First display (triggers loading):")
    print(proxy.display())

    print("\nSecond display (already loaded):")
    print(proxy.display())

    # Demo 2: Protection Proxy
    print("\n\n--- Protection Proxy (Access Control) ---")
    secure_img = SecureImage("confidential.jpg", ["admin", "manager"])

    secure_img.set_user("admin")
    print(secure_img.display())

    print()
    secure_img.set_user("guest")
    print(secure_img.display())

    # Demo 3: Logging Proxy
    print("\n\n--- Logging Proxy ---")
    logging_img = LoggingImageProxy("tracked_image.jpg")

    logging_img.display()
    logging_img.get_info()
    logging_img.display()

    print("\nAccess log:")
    for entry in logging_img.get_access_log():
        print(f"  {entry}")

    # Demo 4: Caching Proxy
    print("\n\n--- Caching Proxy ---")
    CachingImageProxy.clear_cache()

    # Access same file multiple times
    for _ in range(3):
        img = CachingImageProxy("shared_resource.jpg")
        img.display()

    print(f"\nCache size: {CachingImageProxy.get_cache_size()} images")

    # Access different file
    img2 = CachingImageProxy("another_file.jpg")
    img2.display()

    print(f"Cache size: {CachingImageProxy.get_cache_size()} images")

    # Demo 5: Remote Proxy
    print("\n\n--- Remote Proxy ---")
    remote = RemoteImageProxy("banner.png", "images.cdn.com")
    print(remote.display())
    print(f"Info: {remote.get_info()}")

    # Demo 6: Compare proxy with direct access
    print("\n\n--- Comparison: Proxy vs Direct Access ---")

    print("\nWithout proxy (immediate loading):")
    start = time.time()
    direct = HighResolutionImage("direct.jpg")
    direct_time = time.time() - start
    print(f"Time: {direct_time:.3f}s")
    _ = direct.display()  # Use the variable

    print("\nWith proxy (deferred loading):")
    start = time.time()
    proxy = ImageProxy("proxied.jpg")
    proxy_create_time = time.time() - start
    print(f"Proxy creation time: {proxy_create_time:.3f}s")

    start = time.time()
    proxy.display()
    proxy_load_time = time.time() - start
    print(f"First access time: {proxy_load_time:.3f}s")

    print("\n" + "=" * 60)
    print("Types of Proxies:")
    print("=" * 60)
    print("1. Virtual Proxy: Delays expensive object creation")
    print("2. Protection Proxy: Controls access to object")
    print("3. Remote Proxy: Represents object in different address space")
    print("4. Logging Proxy: Adds logging when accessing object")
    print("5. Caching Proxy: Caches results of expensive operations")

    print("\n" + "=" * 60)
    print("Benefits of Proxy Pattern:")
    print("=" * 60)
    print("1. Control access without modifying original object")
    print("2. Manage lifecycle of service objects")
    print("3. Works even if service object isn't ready/available")
    print("4. Open/Closed: new proxies without changing service")


if __name__ == "__main__":
    main()
